web/src/features/column-visibility/hooks/useColumnVisibility.ts

```diff
@@ -4,12 +4,51 @@ import useLocalStorage from "@/src/components/useLocalStorage";
 import { useEffect } from "react";
 import { isEqual } from "lodash";
 
+const readStoredVisibilityState = (
+  localStorageKey: string,
+): VisibilityState => {
+  if (typeof window === "undefined") {
+    return {};
+  }
+  try {
+    const storedValue = localStorage.getItem(localStorageKey);
+    return storedValue ? JSON.parse(storedValue) : {};
+  } catch (error) {
+    console.error("Error reading from local storage", error);
+    return {};
+  }
+};
+
 function useColumnVisibility<TData>(
   localStorageKey: string,
   columns: LangfuseColumnDef<TData>[],
 ) {
   const initialVisibilityState = () => {
     const visibilityState: VisibilityState = {};
+    const storedVisibilityState = readStoredVisibilityState(localStorageKey);
+
+    // With virtual detail columns we must ensure state in local storage is valid given current project data and possibly upsert
+    if (Object.keys(storedVisibilityState).length > 0) {
```

Comment:
do we need to upsert or could we default to the default in case the col is not in the visibility state yet and then upsert once the state changes (user interacting with drop down)